### ISSUE TITLE
Create a new mesh with MFEM by selecting a piece of the main mesh.

### ISSUE THREAD

**Issue by LeilaSalomon** [mfem/mfem#1551]
Hello,

I begin to generate a .msh mesh with gmsh, using  Mesh *mesh = new Mesh(mesh_file, 1, 1). I'm interested to know how can I create an other mesh (with MFEM) using the first mesh. I would like to select just a piece of the main mesh.

Regards,

Leïla

**Comment by mlstowell** [comment]
Hello, @LeilaSalomon ,

Have a look at the [Trimmer](https://github.com/mfem/mfem/blob/master/miniapps/meshing/trimmer.cpp) mini app.  We just added this yesterday if I'm not mistaken.  This miniapp removes elements based on their element attribute numbers.  The basic strategy is to examine the existing mesh and determine the number of elements and, optionally, boundary elements that will be copied to the new mesh.  Allocate the new mesh. Copy vertices, elements, and boundary elements.  Remove any unused vertices. That's about it.

Best wishes,
Mark

**Comment by ben-s-southworth** [comment]
@mlstowell I have a related question I've been wanting to work on soon:

Suppose I have a global serial mesh (standard load mesh and set refine), then construct a parallel mesh and refine as usual. The ParMesh is indexed locally only on the subdomain that the processor owns, and globally indexed w.r.t. other processors.

I want to do a domain-global discretization, assembly, and solve on processor, which uses the refined mesh from the parallel mesh overlaid on the global serial mesh. The catch is I then need to communicate and compute residuals on the parallel fine mesh, which seems related to @LeilaSalomon's question. Does that make sense/would this be feasible (w/o too much work in the guts of MFEM)?

**Comment by mlstowell** [comment]
Hi, @ben-s-southworth ,

I'm not sure I understand your proposal but it sounds like a domain decomposition scheme.  I know @dylan-copeland and @psocratis have been working on this approach for wave propagation problems.  Perhaps it would be useful to compare notes with them.

Best wishes,
Mark

**Comment by dylan-copeland** [comment]
@ben-s-southworth Could you please clarify what you mean by "domain-global discretization"? 

The first paragraph of your comment describes the usual `ParMesh` construction in MFEM, for a single global domain. `ParFiniteElementSpace` then defines the local and global DOF indices. It sounds like you are suggesting a "serial" `FiniteElementSpace` on the entire domain, so there is no concept of local and global DOFs? If so, then the question is how to map between the `FiniteElementSpace` DOFs on the entire domain and the `ParFiniteElementSpace` DOFs on the portion of the domain seen by each process. Is that the question?

**Comment by ben-s-southworth** [comment]
Hi @dylan-copeland, think you're spot on (with the additional question of how do I form the `FiniteElementSpace`). Just to be sure, suppose we have a 1d mesh and 2 processors:
       P1                P2
x        x      |       x        x
Then we form a ParMesh and do 2 Parallel refinements. Then in parallel P1 owns the local mesh
      P1
xxxxxxxx |
and P2 owns
                         P2
                  | xxxxxxxx
I want to form the standard a `ParFiniteElementSpace` using the ParMesh above, and also a "serial" `FiniteElementSpace` on each processor that looks like, e.g.,
      P1
xxxxxxxx |    x    x
So P1 has a `FiniteElementSpace` defined on a global mesh, where this mesh is refined in the parallel home domain of P1 (identical to ParMesh) and also contains coarse elements that cover the entire domain (same structure/coarse mesh on all processors, from initial serial mesh). Then each processor will do a (serial) solve on its own serial  `FiniteElementSpace` and map the result to on the refined/locally owned part of the mesh corresponding to the `ParFiniteElementSpace`.

Does that make sense? It's a little tricky to describe, but is very related to some work Rob Falgout has done called AMG-DD.

**Comment by psocratis** [comment]
@ben-s-southworth,  Is this some kind of a non overlapping Schwarz method? 

I'm kind of lost in the details of your explanation. What do you mean "  P1 has a FiniteElementSpace defined on a global mesh" ? 

Here is what I understood and please correct me if I am wrong...You have a parallel mesh and ParFiniteElementSpace partitioned to a number of processors. You want to create a local submesh and a FiniteElementSpace local to each processor, solve the local problems and communicate solutions/residuals with the "global" ParFiniteElementSpace? 
If that's the case then yes we have something like this (in dev branches) 

Thanks, 

Socratis

**Comment by dylan-copeland** [comment]
My understanding is that you want to do the same global coarse solve on all processes, so that they can be done in serial without communication for the solve. There will need to be a parallel gather for the input coarse vector, but then the solve is serial (global) and the mapping to the local `ParFiniteElementSpace` DOFs can be done locally (without communication). Is that right? That may not be the easiest way to implement the solver, but first we need to have a common understanding of the goal and then find the best/easiest solution.

**Comment by LeilaSalomon** [comment]
> Hello, @LeilaSalomon ,
> 
> Have a look at the [Trimmer](https://github.com/mfem/mfem/blob/master/miniapps/meshing/trimmer.cpp) mini app. We just added this yesterday if I'm not mistaken. This miniapp removes elements based on their element attribute numbers. The basic strategy is to examine the existing mesh and determine the number of elements and, optionally, boundary elements that will be copied to the new mesh. Allocate the new mesh. Copy vertices, elements, and boundary elements. Remove any unused vertices. That's about it.
> 
> Best wishes,
> Mark

Thank you, but is it possible to create a new mesh which will be smaller than the original mesh? With a different number of vertices ?

**Comment by ben-s-southworth** [comment]
Hi @psocratis and @dylan-copeland,
It is more of a asynchronous Schwarz-like method, where each iteration there is an MPI gather as you mention @dylan-copeland. Let me try some plots. Suppose here is a ParMesh split over 4 processors:
[ParMesh.pdf](https://github.com/mfem/mfem/files/4793201/ParMesh.pdf)
Each of these processors also owns a serial Mesh of the following form(s):
[SerMesh.pdf](https://github.com/mfem/mfem/files/4793204/SerMesh.pdf)
In a given iteration, each processor solves the sequential problem on the above serial mesh, and then the updated solution/residual on the ParMesh is taken by value from appropriate processor's serial mesh (e.g., points in bottom left quadrant com from P1, bottom right from P2, etc.).

**Comment by mlstowell** [comment]
Hi, @LeilaSalomon ,

Yes, the new mesh will have fewer vertices than the original mesh.  There are at least two ways to do this.  MFEM has a method, `Mesh::RemoveUnusedVertices` which will determine which vertices do not belong to any element or boundary element.  It then removes them from the list of vertices and renumbers the element and boundary element connectivities.  This is convenient but it can lead to a large number of temporary vertices.  The alternative is to determine which vertices should be kept during the mesh examination stage.  Then renumber the connectivities as the elements and boundary elements are copied.

Best wishes,
Mark

**Comment by dylan-copeland** [comment]
Hi @ben-s-southworth,

I think this is feasible in MFEM but not easy, as MFEM was not designed for such complicated mappings of spaces. On a high level, the algorithm seems to be as follows.

1. Generate a coarse global serial mesh that is the same on all processes (done as usual), and construct a `FiniteElementSpace` on it. This step is trivial, as in ex1/ex1p.
2. Construct a `ParMesh` and do parallel refinements, then construct a `ParFiniteElementSpace`. This is trivial, but you will need to save prolongation operators from the coarse to fine spaces at each refinement, as in geometric multigrid. I don't know of standard examples in MFEM for this, but examples/solvers-dev/maxwellp_real.cpp in branch gmg-solver is one example.
3. You will also need a map between the `FiniteElementSpace` and the `ParFiniteElementSpace`, which is tricky (not supported directly in MFEM). I have code in a branch that maps DOFs between two spaces, and I can point you to it with some details, but I don't recommend it. Basically, you first need a map between elements in the two meshes, and then use `GetElementDofs` in the two spaces to get their respective sets of element DOFs, and then determine the map between them. For quad/hex meshes, this is not too bad as DOFs can be mapped bijectively with a possibly sign change, but for simplices there can be a linear map due to different orderings, and I have only done the hexahedral case.

I think a simpler way is to do steps 1 and 2 above, and gather the coarse HypreParMatrix on all processes into a SparseMatrix. This is easier, because the matrices are simpler data structures than meshes and finite element spaces. Each process stores entire global rows of the HypreParMatrix, which you can simply gather (via MPI) to get the	I, J, data vectors to construct a SparseMatrix. For example, see `GatherHypreParMatrix` in examples/ddoper.cpp in branch complex-strumpack-dev-dylan. It is easier to map between the global, serial SparseMatrix and the local, parallel HypreParMatrix indices than to map DOFs.

This is diverging from the original issue, so maybe we should create a separate issue if this needs to be discussed in much more detail.

**Comment by psocratis** [comment]
> Hi @psocratis and @dylan-copeland,
> It is more of a asynchronous Schwarz-like method, where each iteration there is an MPI gather as you mention @dylan-copeland. Let me try some plots. Suppose here is a ParMesh split over 4 processors:
> [ParMesh.pdf](https://github.com/mfem/mfem/files/4793201/ParMesh.pdf)
> Each of these processors also owns a serial Mesh of the following form(s):
> [SerMesh.pdf](https://github.com/mfem/mfem/files/4793204/SerMesh.pdf)
> In a given iteration, each processor solves the sequential problem on the above serial mesh, and then the updated solution/residual on the ParMesh is taken by value from appropriate processor's serial mesh (e.g., points in bottom left quadrant com from P1, bottom right from P2, etc.).

@ben-s-southworth 
Ok..it's clear now what you need. I agree with Dylan that this would not be trivial. Let me ask a couple of questions, based on @dylan-copeland comments too.
1) What meshes are you considering ? 2D/3D, quad/hex or more general ? 
2) Are the refinements only uniform or are you thinking AMR? 
3) I am not sure that the prolongation operator from coarse to fine mesh is needed, since the coarse dofs will be used only for the "local-serial" solve and then only the "fine" dofs will contribute to the "global-parallel" system. Is that right @ben-s-southworth ? 

Here are my thoughts: (say uniform refinements and quad mesh)
1) Construct the ParFiniteElementSpace  on the global pmesh (On each processor you still have the serial mesh)
2) Perform refinement on the pmesh. At this point you will have to mark the elements on each processor that are refined (as elements on pmesh) and manually refine them locally (as elements on the local mesh) 
3) constructing the map between the mesh and the pmesh on each processor is not hard now using the local (updated after refinement) FiniteElementSpace and the ParFiniteElementSpace.
4) One possible issue that I see is that the serial meshes will be non-conforming any more. So that might need some careful handling  

Does this make sense ? 

Socratis

**Comment by dylan-copeland** [comment]
@psocratis It seems you are thinking that the serial `FiniteElementSpace` is local (on a subdomain), while I am thinking it is a global coarse space on the entire domain. @ben-s-southworth which did you intend?

**Comment by psocratis** [comment]
> @psocratis It seems you are thinking that the serial `FiniteElementSpace` is local (on a subdomain) with local solves, while I am thinking it is a global coarse space on the entire domain with a global solve (identical on all processes). @ben-s-southworth which did you intend?

No...I got that point. @ben-s-southworth said "points in bottom left quadrant com from P1, bottom right from P2, etc..". So the coarse dofs will not contribute anywhere except the local-serial solve.

**Comment by LeilaSalomon** [comment]
> Hi, @LeilaSalomon ,
> 
> Yes, the new mesh will have fewer vertices than the original mesh. There are at least two ways to do this. MFEM has a method, `Mesh::RemoveUnusedVertices` which will determine which vertices do not belong to any element or boundary element. It then removes them from the list of vertices and renumbers the element and boundary element connectivities. This is convenient but it can lead to a large number of temporary vertices. The alternative is to determine which vertices should be kept during the mesh examination stage. Then renumber the connectivities as the elements and boundary elements are copied.
> 
> Best wishes,
> Mark

Thank you very much Mark, it seems to work very well !!!

**Comment by LeilaSalomon** [comment]
> > Hi, @LeilaSalomon ,
> > Yes, the new mesh will have fewer vertices than the original mesh. There are at least two ways to do this. MFEM has a method, `Mesh::RemoveUnusedVertices` which will determine which vertices do not belong to any element or boundary element. It then removes them from the list of vertices and renumbers the element and boundary element connectivities. This is convenient but it can lead to a large number of temporary vertices. The alternative is to determine which vertices should be kept during the mesh examination stage. Then renumber the connectivities as the elements and boundary elements are copied.
> > Best wishes,
> > Mark
> 
> Thank you very much Mark, it seems to work very well !!!

Futhermore, is it possible to create new boundary elements with a 1D mesh ? In this exemple, we use GetNFaces and FaceElement, but it doesn't exist in one dimension no ?

**Comment by v-dobrev** [comment]
You probably want `Mesh::GetNumFaces` -- this is the method that returns the number of `(dim-1)`-dimensional mesh entities which for 1D mesh are the vertices.

The face elements (i.e. the `(dim-1)`-dimensional mesh entities) are not generated in 1D (they will be simply "point" elements) since they are typically not needed. If you really need them you can uncomment this line: https://github.com/mfem/mfem/blob/bf3423dcbac18f7e3b0e8a263435d466f4687c5a/mesh/mesh.cpp#L5037

**Comment by ben-s-southworth** [comment]
@psocratis I think you are correct - each processor does a domain-global serial solve, e.g., on the meshes shown for P1, ..., P4 I posted above. But (i) the coarse DOFs (everything outside of what is "owned" in parallel) don't contribute to anything in parallel, and (ii) the "fine" DOFs are identical on the serial mesh and parallel mesh. @dylan-copeland, I think maybe this was not clear - the mesh that the serial solve is performed on is different for every processor, because it is refined exactly as the ParMesh where this processors owns elements in parallel (e.g., compare P1 and P2 plots in SerMesh.pdf).

Re other questions - this for uniform refinement and 2d/3d meshes, but I am not so picky on element type. It is more to test and demonstrate this algorithm. Yes, only the "fine" DOFs contribute to the parallel system. The serial mesh will need be nonconforming, but I will deal with that once I get the basic meshing and serial/parallel interface worked out.

So let me run a slightly more specific set of steps by you. Creating and refining the ParMesh is easy. Then to manually refine the serial mesh, I could use Mesh.GeneratePartitioning to determine the element indices owned by *this processor (partitioning[i] == rank), then refine these elements the same number of times as done for the ParMesh. This will all be deterministic right, so the ordering of refined elements in the ParMesh will match the locally refined group of elements in the serial mesh? Before/during this process, is it possible to use the attribute feature (or something else?) to mark elements of the serial mesh that are "fine" vs "coarse"? If so, it would be easy to then (after refinement) make my own mapping from coarse to fine by, e.g., storing the indices of all elements marked "fine" and setting/extracting the subvector of those DOFs to map between the serial mesh and ParMesh.

**Comment by mlstowell** [comment]
Hi, @LeilaSalomon,

While looking into your 1D support question I noticed that the pre-existing boundary elements don't have the correct attribute numbers in the new mesh.  I'll try to submit a bug fix soon.

Best wishes,
Mark

**Comment by mlstowell** [comment]
Hi, @LeilaSalomon ,

The bug fix and 1D support have been added in PR #1558.

Best wishes,
Mark

**Comment by psocratis** [comment]
Hi @ben-s-southworth , 

I agree with most of the steps you described. One thing I am not sure is on the ordering of the refined elements. This shouldn't really be a problem though since you already know the map between local mesh and global pmesh for the coarse elements. After the refinement you can can find the coarse "parent" element of a fine element using 

https://github.com/mfem/mfem/blob/bf3423dcbac18f7e3b0e8a263435d466f4687c5a/mesh/ncmesh.hpp#L46-L76

I think yes you can use the element attributes for marking or I guess you can create your own lists. I still don't think handling the hanging nodes will be trivial though. I am not saying it cannot be done but it might require some thinking. Think of a fine dof on the pmesh that is a constrained dof on the "local-serial" problem. That dof is not a TrueDof... so.. after solving the "local-serial" problem you would have to transfer from TrueDofs to dofs (element dofs) and then from there you can transfer to the global "fine dofs".

I hope this helps. 

Socratis

**Comment by ben-s-southworth** [comment]
Hi @psocratis, thanks for the help. I don't quite follow the second part, which is likely just a lack of understanding of a non-conforming mesh/discretization. On the serial mesh at the interface between my "fine" and "coarse" elements, there will be hanging nodes. However, I thought these hanging nodes would still be TrueDOFs/need to be solved for, as they still have a right-hand side and interdependent on other conforming DOFs. Is that not correct?

**Comment by psocratis** [comment]
Hi @ben-s-southworth , 

The answer in no, a hanging node does not enter the linear system (i.e, is not a TrueDof). A hanging (or constrained) node is constrained by its "parents". For example look at the picture below (taken from the MFEM paper from https://arxiv.org/pdf/1911.09220.pdf) 

![image](https://user-images.githubusercontent.com/14336752/85435397-3def9c80-b53c-11ea-84c8-1f3994ad6039.png)

 Dof c is constrained by dofs a and b (written as a linear combination of a and b). In this case only a and b enter the linear system and through them you can post-process and get the value of c. All this is done by MFEM of course using the prolongation operator. 

I advise you take a look at the paper above (section 7.2) and we can continue the discussion (maybe in private if preferred because we are going a bit off topic). 

Thanks, 

Socratis

**Comment by ben-s-southworth** [comment]
Excellent, thanks! This is good for now, I am happy for the issue to be closed, but will defer to @LeilaSalomon given this was originally their post.
